### PR TITLE
Implement reader for Syslog RFC5424

### PR DESCRIPTION
**Pull request by knapperzbusch**
This PR adds a Syslog parser and reader. With this, it is possible to import Syslog messages in the format defined in [RFC5424](https://tools.ietf.org/html/rfc5424). If a message can't be parsed, the message will be handled as single string.

### CODE REVIEW

**Comment by dominiklohmann** [comment]
[ch9860]

File: libvast/vast/concept/parseable/core/repeat.hpp [hunk 1]
```diff
@@ -88,8 +85,8 @@ class dynamic_repeat_parser
 
 private:
   Parser parser_;
-  const T& min_;
-  const U& max_;
+  const T min_;
+  const T max_;
```
Comment (dominiklohmann) [inline libvast/vast/concept/parseable/core/repeat.hpp]: Just a heads up: You can likely revert this change tomorrow, as I plan to fix this (and the same bug in the skip parser) upstream.

File: libvast_test/src/main.cpp [hunk 2]
```diff
@@ -40,27 +40,32 @@ const char* ssl = VAST_TEST_PATH "artifacts/logs/zeek/ssl.log";
...
 namespace bgpdump {
 
-const char* updates20140821 = VAST_TEST_PATH
-  "artifacts/logs/bgpdump/updates20140821.txt";
-const char* updates20180124 = VAST_TEST_PATH
-  "artifacts/logs/bgpdump/updates20180124.txt";
+const char* updates20140821
+  = VAST_TEST_PATH "artifacts/logs/bgpdump/updates20140821.txt";
+const char* updates20180124
+  = VAST_TEST_PATH "artifacts/logs/bgpdump/updates20180124.txt";
```
Comment (dominiklohmann) [inline libvast_test/src/main.cpp]: These formatting changes seem unrelated.

File: libvast_test/src/main.cpp [hunk 3]
```diff
@@ -88,9 +93,11 @@ int main(int argc, char** argv) {
...
   }
   if (start != argc) {
-    auto res = caf::message_builder(argv + start, argv + argc).extract_opts({
-      {"gperftools,g", "enable gperftools profiler for actor tests"},
-    });
+    auto res
+      = caf::message_builder(argv + start, argv + argc)
+          .extract_opts({
+            {"gperftools,g", "enable gperftools profiler for actor tests"},
+          });
```
Comment (dominiklohmann) [inline libvast_test/src/main.cpp]: This formatting change seems unrelated.

File: libvast_test/vast/test/data.hpp [hunk 4]
```diff
@@ -28,6 +28,10 @@ extern const char* bview;
 extern const char* updates20150505;
 
 } // namespace mrt
+
+namespace syslog {
+extern const char* syslog_msgs;
+}
```
Comment (dominiklohmann) [inline libvast_test/vast/test/data.hpp]:
```suggestion

namespace syslog {

extern const char* syslog_msgs;

} // namespace syslog
```

File: libvast/vast/format/syslog.hpp [hunk 5]
```diff
@@ -0,0 +1,299 @@
...
+  using value_type = typename std::decay_t<Parser>::attribute;
+  using attribute = std::conditional_t<detail::is_container<value_type>,
+                                       value_type, caf::optional<value_type>>;
+
+  explicit maybe_nil_parser(Parser parser) : parser_{std::move(parser)} {
+  }
+
+  template <class Iterator, class Attribute>
+  auto parse(Iterator& f, const Iterator& l, Attribute& x) const {
+    using namespace parsers;
```
Comment (dominiklohmann) [inline libvast/vast/format/syslog.hpp]: This parser does not appear to be using the namespace parsers.

File: libvast/vast/format/syslog.hpp [hunk 6]
```diff
@@ -0,0 +1,299 @@
...
+#include <caf/sum_type.hpp>
+
+#include <string>
+#include <type_traits>
+#include <utility>
+
+namespace vast::format::syslog {
+
+// This namespace includes the syslog message defintion of the syslog protocoll
+// defined in [RFC 5424](https://tools.ietf.org/html/rfc5424).
```
[the remark this answers is not in the document]
Comment (dominiklohmann) [reply libvast/vast/format/syslog.hpp]: Generally we try to format our documentation for Doxygen, and try to use full sentences. This applies to all comments in this PR.

File: libvast/vast/format/syslog.hpp [hunk 7]
```diff
@@ -0,0 +1,299 @@
...
+    auto p = pri >> version >> ' ' >> timestamp >> ' ' >> hostname >> ' '
+             >> app_name >> ' ' >> process_id >> ' ' >> msg_id;
+
+    if constexpr (std::is_same_v<Attribute, unused_type>)
+      return p(f, l, unused);
+    else
+      return p(f, l, x.version, x.ts, x.hostname, x.app_name, x.process_id,
+               x.msg_id);
+  }
+}; // namespace vast::format::syslog
```
Comment (dominiklohmann) [inline libvast/vast/format/syslog.hpp]:
```suggestion
};
```

File: libvast/vast/format/syslog.hpp [hunk 8]
```diff
@@ -0,0 +1,296 @@
...
+    auto is_version = [](uint16_t in) { return in > 0; };
+    auto version = integral_parser<uint16_t, 3>{}.with(is_version);
+    auto hostname = maybe_nil(rep(print - ' ', 1, 255));
+    auto app_name = maybe_nil(rep(print - ' ', 1, 48));
+    auto process_id = maybe_nil(rep(print - ' ', 1, 128));
+    auto msg_id = maybe_nil(rep(print - ' ', 1, 32));
+    auto timestamp = maybe_nil(parsers::time);
+    auto p = pri >> version >> ' ' >> timestamp >> ' ' >> hostname >> ' '
+             >> app_name >> ' ' >> process_id >> ' ' >> msg_id;
+
```
Comment (dominiklohmann) [inline libvast/vast/format/syslog.hpp]:
```suggestion
```

File: libvast/vast/format/syslog.hpp [hunk 9]
```diff
@@ -0,0 +1,296 @@
...
+using parameter = std::tuple<std::string, std::string>;
+
+// The parameter_parser parses one structured data paramter
+struct parameter_parser : parser<parameter_parser> {
+  using attribute = parameter;
+
+  template <class Iterator, class Attribute>
+  bool parse(Iterator& f, const Iterator& l, Attribute& x) const {
+    using parsers::print, parsers::rep, parsers::ch;
+
```
Comment (dominiklohmann) [inline libvast/vast/format/syslog.hpp]:
```suggestion
```

File: libvast/vast/format/syslog.hpp [hunk 10]
```diff
@@ -0,0 +1,296 @@
...
+    // space, =, ", and ] are not allowed in the key of the parameter
+    auto key = rep(print - '=' - ' ' - ']' - '"', 1, 32);
+    // \ is used to escape specific character
+    auto esc = ignore(ch<'\\'>);
+    // ], ", \ need to be escaped
+    auto escaped = esc >> (ch<']'> | ch<'\\'> | ch<'"'>);
+    auto value = escaped | (print - ']' - '"' - '\\');
+    auto p = ' ' >> key >> '=' >> '"' >> *value >> '"';
+
```
Comment (dominiklohmann) [inline libvast/vast/format/syslog.hpp]:
```suggestion
```

File: libvast/vast/format/syslog.hpp [hunk 11]
```diff
@@ -0,0 +1,296 @@
...
+// The structured_data_element_parser parses one structured data element
+struct structured_data__element_parser
+  : parser<structured_data__element_parser> {
+  using attribute = structured_data_element;
+
+  template <class Iterator, class Attribute>
+  bool parse(Iterator& f, const Iterator& l, Attribute& x) const {
+    using parsers::print, parsers::rep;
+
```
Comment (dominiklohmann) [inline libvast/vast/format/syslog.hpp]:
```suggestion
```

File: libvast/vast/format/syslog.hpp [hunk 12]
```diff
@@ -0,0 +1,296 @@
...
+// The structured_data_parser parses the structured_data elements of one Syslog
+// message
+struct structured_data_parser : parser<structured_data_parser> {
+  using attribute = structured_data;
+
+  template <class Iterator, class Attribute>
+  bool parse(Iterator& f, const Iterator& l, Attribute& x) const {
+    using namespace parsers;
+
```
Comment (dominiklohmann) [inline libvast/vast/format/syslog.hpp]:
```suggestion
```

File: libvast/vast/format/syslog.hpp [hunk 13]
```diff
@@ -0,0 +1,296 @@
...
+// The message_content_parser parses the message component of a Syslog message
+struct message_content_parser : parser<message_content_parser> {
+  using attribute = message_content;
+  template <class Iterator, class Attribute>
+  bool parse(Iterator& f, const Iterator& l, Attribute& x) const {
+    using namespace parsers;
+    using namespace parser_literals;
+    auto bom = "\xEF\xBB\xBF"_p;
+    auto p = (bom >> +any) | (+any) | eoi;
+
```
Comment (dominiklohmann) [inline libvast/vast/format/syslog.hpp]:
```suggestion
```

File: libvast/vast/format/syslog.hpp [hunk 14]
```diff
@@ -0,0 +1,296 @@
...
+// The message_content_parser parses the message component of a Syslog message
+struct message_content_parser : parser<message_content_parser> {
+  using attribute = message_content;
+  template <class Iterator, class Attribute>
+  bool parse(Iterator& f, const Iterator& l, Attribute& x) const {
+    using namespace parsers;
+    using namespace parser_literals;
+    auto bom = "\xEF\xBB\xBF"_p;
+    auto p = (bom >> +any) | (+any) | eoi;
```
Comment (dominiklohmann) [inline libvast/vast/format/syslog.hpp]:
```suggestion
    auto p = (bom >> +any) | +any | eoi;
```

File: libvast_test/src/main.cpp [hunk 15]
```diff
@@ -40,27 +40,32 @@ const char* ssl = VAST_TEST_PATH "artifacts/logs/zeek/ssl.log";
...
-  "artifacts/logs/mrt/updates20150505.0";
+const char* updates20150505
+  = VAST_TEST_PATH "artifacts/logs/mrt/updates20150505.0";
 
 } // namespace mrt
+
+namespace syslog {
+const char* syslog_msgs
+  = VAST_TEST_PATH "artifacts/logs/syslog/syslog-test.txt";
+} // namespace syslog
```
Comment (dominiklohmann) [inline libvast_test/src/main.cpp]:
```suggestion
namespace syslog {

const char* syslog_msgs
  = VAST_TEST_PATH "artifacts/logs/syslog/syslog-test.txt";
  
} // namespace syslog

```

File: libvast/vast/format/syslog.hpp [hunk 16]
```diff
@@ -0,0 +1,296 @@
...
+
+// header, data and msg describe one syslog message
+struct message {
+  header header;
+  structured_data data;
+  caf::optional<message_content> msg;
+};
+
+// the message parses uses the header_parser, structured_datas_parser and the
+// message_content_parser to parse one Syslog message
```
Comment (dominiklohmann) [inline libvast/vast/format/syslog.hpp]:
```suggestion
/// Parser for Syslog messages.
/// @relates message
```

File: libvast/vast/format/syslog.hpp [hunk 17]
```diff
@@ -0,0 +1,296 @@
...
+    using namespace parsers;
+    using namespace parser_literals;
+    auto bom = "\xEF\xBB\xBF"_p;
+    auto p = (bom >> +any) | (+any) | eoi;
+
+    return p(f, l, x);
+  }
+};
+
+// header, data and msg describe one syslog message
```
Comment (dominiklohmann) [inline libvast/vast/format/syslog.hpp]:
```suggestion
/// A Syslog message.
```

File: libvast/vast/format/syslog.hpp (hunk 16 above)
Comment (dominiklohmann) [reply libvast/vast/format/syslog.hpp]: You can apply this style of documentation comment to all other classes in this file as well.

File: libvast/vast/format/syslog.hpp [hunk 18]
```diff
@@ -0,0 +1,296 @@
...
+  caf::error
+  read_impl(size_t max_events, size_t max_slice_size, consumer& f) override;
+
+private:
+  std::unique_ptr<std::istream> input_;
+  std::unique_ptr<detail::line_range> lines_;
+  type syslog_rfc5424_type_;
+  type syslog_unkown_type_;
+};
+}; // namespace vast::format::syslog
```
Comment (dominiklohmann) [inline libvast/vast/format/syslog.hpp]:
```suggestion
};

} // namespace vast::format::syslog
```

File: CHANGELOG.md [hunk 19]
```diff
@@ -10,6 +10,10 @@ Every entry has a category for which we use the following visual abbreviations:
 
 ## [Unreleased]
 
+-  🎁 The new `vast import syslog` command allows importing Syslog messages
+   as defined in [RFC5424](https://tools.ietf.org/html/rfc5424).
+   [#770](https://github.com/knapperzbusch/vast/pull/770)
```
Comment (dominiklohmann) [inline CHANGELOG.md]:
```suggestion
- 🎁 The new `vast import syslog` command allows importing Syslog messages
  as defined in [RFC5424](https://tools.ietf.org/html/rfc5424).
  [#770](https://github.com/knapperzbusch/vast/pull/770)
```

File: doc/cli/vast-import-syslog.md [hunk 20]
```diff
@@ -0,0 +1,6 @@
+The syslog import format uses [RFC5424](https://tools.ietf.org/html/rfc5424) consumes syslog messages in format defined in RFC 5424. 
+
+In order to run consume syslog messages use the following command:
+
+```
+vast import syslog -r <path-to-syslog-file>
```
Comment (dominiklohmann) [inline doc/cli/vast-import-syslog.md]:
```suggestion
Ingest Syslog messages into VAST. The following formats are supported:

- [RFC 5424](https://tools.ietf.org/html/rfc5424)
- A fallback format that consists only of the Syslog message.

If you require support for any other Syslog formats, feel free to contact us at
info@tenzir.com.

Usage example:

```sh
# Import from file.
vast import syslog -r path/to/sys.log

# Continuously import from a stream.
syslog | vast import syslog
```
```

Comment (knapperzbusch) [reply doc/cli/vast-import-syslog.md]: I have a question related to this suggestion. Shall I mention that the command `syslog | vast import syslog` interprets each message as one string for now?

File: libvast/vast/concept/parseable/core/repeat.hpp (hunk 1 above)
Comment (dominiklohmann) [reply libvast/vast/concept/parseable/core/repeat.hpp]: This has been fixed upstream.

**Comment by dominiklohmann** [comment]
@knapperzbusch should this still be in draft mode?

**Comment by knapperzbusch** [comment]
I will do rebase with master today night or tomorrow morning, to resolve the conflicts.

**Comment by dominiklohmann** [comment]
Sounds good. Will do one last review pass then (likely next week), so we can get this merged.

**Comment by dominiklohmann** [comment]
I took the freedom to rebase this once more and rewrite some comments—no more code changes were necessary.

Thanks for your contribution, @knapperzbusch! We'll merge once CI has run through successfully.
